content/blog/infrastructure-as-code-with-aws-cdk.md

```diff
@@ -0,0 +1,120 @@
+
+### The Requirements
+
+* Deploy applications on AWS ECS
+* Use RDS PostgreSQL as the database
+* Ensure database isolation for each client
+* Deploy the same application multiple times for different clients
+* Minimize onboarding effort for each new client
+* Manage multiple environments (lower and higher) with similar infrastructure components, but with configuration changes
+
+### The Challenge
+
+To illustrate the challenge, let's consider an example. Suppose we have a bill generator service that needs to be deployed for multiple clients. The application code remains the same, but each client requires its own isolated database. We could either deploy the application on our own AWS infrastructure or on the client's AWS environment. However, this would require manual effort and configuration for each new client, which wasn't scalable.
+
+### The Solution: AWS CDK to the Rescue
+
+* Define our infrastructure as code, making it easy to manage and version-control
+* Create a one-click solution for onboarding new clients, minimizing manual effort
+* Manage multiple environments with similar infrastructure components, but with configuration changes
+* Package our CI/CD pipeline, application deployment, configuration management, and container image management into a single CDK codebase
+
+## Key Abilities of AWS CDK
+
+1. **Sensible Defaults**: Reduces the number of code lines we need to write, making it easier to define infrastructure.
+2. **Reusable Libraries**: Allows us to build reusable libraries, reducing the amount of copy-pasting and making it easier to manage complex infrastructure.
+3. **Automated Rollback**: Automatically rolls back resource creation if it fails, ensuring that our infrastructure is always in a consistent state.
+4. **Event Logging**: Shows the events of logs while resource creation, providing visibility into the deployment process.
+5. **Single-Stack Deployment**: Useful for application and service deployments, CDK uses CloudFormation under the hood to deploy a single Stack, reducing conflicts and parallel deployments per environment.
+
+## Important Considerations Before Choosing AWS CDK
+
+1. **Provider Limitations**: We can't switch providers within the same application. This means that if we’re deploying our application to a single AWS account and region, we can't create resources in another region or account in the same run.
+2. **No Cross-Account or Cross-Region Support**: Unlike Terraform, CDK doesn't support assuming roles or accessing resources from other accounts or regions.
+3. **Unstable Development Stage**: Many CDK services are still in the unstable development stage, which means we can expect breaking changes with every update.
+4. **Limited CDK Modules**: Some key AWS services still don't have CDK modules available. This may force us to use a hybrid approach, combining CDK with other tools.
```

Comment:
Can you give names of services that are missing to give people some sense of what they might not be able to manage using AWS CDK?